include/graphblas/algorithms/ising_machine_sb.hpp

```diff
@@ -0,0 +1,332 @@
+				return rc;
+			}
+#ifdef DEBUG_IMSB
+			// for debugging purposes, print sumJ2
+			std::cout << "sumJ2: " << sumJ2 << '\n';
+#endif
+
+			// rewrite this to use graphblas language
+			rc = rc ? rc : grb::foldl< descr_dense >( sumJ2, static_cast<IOType>( N - 1 ), divide );
+			IOType xi = 0.5;
+			sumJ2 = sqrtX( sumJ2 );
+			rc = rc ? rc : grb::foldl< descr_dense >( xi, sumJ2, divide );
+#ifdef DEBUG_IMSB
+			// for debugging purposes, print xi
+			std::cout << "xi: " << xi << '\n';
+#endif
+
+			/* ---- iteration variables ---- */
+			IOType ps  = p_init;
+			const IOType dps = ( p_end - p_init ) / static_cast<IOType>( num_iters - 1 );
+			// assert len of energies == N
+			assert( energies.size() == num_iters );
+
+			for ( std::size_t iter = 0; iter < num_iters; ++iter ) {
+
+			    /* y_comp += ((-1+ps)*x_comp + xi*(Jx + h)) * dt */
+
+			    // Jx ← J * x_comp
+				grb::set( Jx, ring.template getZero< IOType >() );
+				rc = rc ? rc : grb::mxv< descr_dense >( Jx, J, x_comp, ring );
+				assert( rc == grb::SUCCESS );
+#ifdef DEBUG_IMSB
+				vector_print( Jx, "Jx" );
+#endif
+
+			    // temp ← Jx + h
+				grb::set( temp, ring.template getZero< IOType >() );
+			    rc = rc ? rc : grb::eWiseApply< descr_dense >(
+			        temp, Jx, h, ring.getAdditiveMonoid()
+			    );
+				assert( rc == grb::SUCCESS );
+#ifdef DEBUG_IMSB
+				vector_print( temp, "temp = Jx + h" );
+#endif
+
+			    // temp ← xi * temp
+			    rc = rc ? rc : grb::foldl< descr_dense >( 
+					temp, xi, ring.getMultiplicativeMonoid() 
+				);
+				assert( rc == grb::SUCCESS );
+#ifdef DEBUG_IMSB
+				vector_print( temp, "xi * temp" );
+#endif
+
+
+			    // temp ← temp + (-1+ps) * x_comp
+			    const IOType scale = -1.0 + ps;
+				rc = rc ? rc : grb::eWiseMul< descr_dense >( temp, scale, x_comp, ring );
+				assert( rc == grb::SUCCESS );
+#ifdef DEBUG_IMSB
+				std::cout << "scale: " << scale << '\n';
+				vector_print( temp, "temp + (-1+ps) * x_comp" );
+#endif
+
+			    // y_comp += dt * temp
+			    rc = rc ? rc : grb::eWiseMul< descr_dense >( y_comp, dt, temp, ring );
+				assert( rc == grb::SUCCESS );
+#ifdef DEBUG_IMSB
+				vector_print( y_comp, "y_comp (a)" );
+#endif
+
+			    /* x_comp += dt * y_comp */
+			    rc = rc ? rc : grb::eWiseMul< descr_dense >( x_comp, dt, y_comp, ring );
+				assert( rc == grb::SUCCESS );
+#ifdef DEBUG_IMSB
+				vector_print( x_comp, "x_comp" );
+#endif
+
+			    /* y_comp[ |x|>1 ] = 0 */
+				// mask = np.abs(x_comp) > 1
+			    rc = rc ? rc : grb::eWiseLambda< descr_dense >( [&mask, &x_comp]( const size_t i ) {
+					(void) i;
+					// rewrite this to use graphblas language
+					mask[i] = std::abs(x_comp[i]) > 1;
+					}, mask, x_comp
+				);
+				assert( rc == grb::SUCCESS );
+#ifdef DEBUG_IMSB
+				vector_print( mask, "mask" );
+#endif
+				rc = rc ? rc : grb::foldl< descr_dense >(  
+					y_comp, mask, ring.template getZero< IOType >(), 
+					grb::operators::right_assign<bool,IOType,IOType>()
+				);
```

Comment:
@anyzelman It is a bit unclear how this works. There might be a bug here, since if I use non-zero value to assign to the output, the result if different, perhaps downcasted ?  (also with  right_assign_if)

Reply:
bool -> IOType